src/kcas.ml

```diff
@@ -218,49 +336,67 @@ module Loc = struct
   type 'a t = 'a loc
 
   let make after =
-    { state = Atomic.make @@ new_state after; id = Id.get_unique () }
+    let state = Atomic.make @@ new_state after and id = Id.get_unique () in
+    { state; id }
 
   let get_id loc = loc.id [@@inline]
+  let get loc = eval (Atomic.get loc.state)
 
-  let get loc =
-    let state = Atomic.get loc.state in
-    let before = state.before and after = state.after in
-    if before == after || is_after state.casn then after else before
+  let rec get_as ?(await = `Cancelable) f loc =
+    let before = get loc in
+    match f before with
+    | value -> value
+    | exception Retry.Later ->
+        block await loc before;
+        get_as f loc
 
   let compare_and_set loc before after =
     let state = new_state after in
     cas loc before state
 
-  let update ?(backoff = Backoff.default) loc f =
-    let state' = Atomic.get loc.state in
+  let update ?(await = `Cancelable) ?(backoff = Backoff.default) loc f =
+    let state' = fenceless_get loc.state in
     let before = eval state' in
-    let state = new_state (f before) in
-    if Atomic.compare_and_set loc.state state' state then before
-    else
-      let backoff = Backoff.once backoff in
-      update_no_alloc backoff loc state @@ fun state before ->
-      state.after <- f before
-
-  let modify ?backoff loc f = update ?backoff loc f |> ignore [@@inline]
+    match f before with
+    | after ->
+        if before == after then before
```

Comment:
Yes, this is an optimization to avoid unnecessarily updating locations.  It wasn't implemented before, but with the addition of awaiters it potentially avoids waking up awaiters unnecessarily (as nothing logically changed), which could avoid a lot of unnecessary computation.